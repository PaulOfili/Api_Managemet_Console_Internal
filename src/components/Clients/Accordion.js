import React from 'react';
import { Form, List, Radio } from 'semantic-ui-react';

function Accordion({ array, handleRadio, approve, type}){

    return (
        <div className="accordion mt-3" id="accordionExample">
            {
                array.map((product, index) => (
                <div className="" key={index}>
                    <div className="card-header" id={`heading${product.name}`}>
                        <h2 className="mb-0">
                            <button className="btn btn-link field" type="button" data-toggle="collapse" data-target={`#${type}_${product.name}-${index}`} aria-expanded="true" aria-controls="collapseOne">
                                <label className='product_name'>{product.name}</label>
                            </button>
                        </h2>
                    </div>

                    <div id={`${type}_${product.name}-${index}`} className="collapse" aria-labelledby={`heading${product.name}`} data-parent="#accordionExample">
                        <div className="card-body">
                            { 
                                <List>
                                    {
                                        product.resources.map((resource, resource_index) => (
                                            <List.Item id={`${type}_${product.name}-${index}-${resource_index}`} key={resource_index} style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-between'}}>
                                                <List.Content style={{ width: '50%'}}>
                                                    <List.Header>{resource.name}</List.Header>
                                                </List.Content>
                                                {
                                                approve &&
                                                <List.Content style={{ width: '50%'}}>
                                                    <Form.Group inline style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-around'}}>
                                                        <Form.Field>
                                                            <Radio
                                                                label='Approve'
                                                                name={resource.id}
                                                                value='approve'
                                                                checked={approve[resource.id] === 'approve'}
                                                                onChange={handleRadio}
                                                            />
                                                        </Form.Field>
                                                        <Form.Field>
                                                            <Radio
                                                                label='Decline'
                                                                name={resource.id}
                                                                value='decline'
                                                                checked={approve[resource.id] === 'decline'}
                                                                onChange={handleRadio}
                                                            />
                                                        </Form.Field>
                                                    </Form.Group>
                                                </List.Content>
                                                }
                                            </List.Item>
                                        ))
                                    }
                                </List>
                            }
                        </div>
                    </div>
                </div>
            ))
            }
        </div>
    );
}

export default Accordion;