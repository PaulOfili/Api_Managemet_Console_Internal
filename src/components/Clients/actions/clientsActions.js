import types from '../clientsConstants';

export const getClients = () => {
    return {
        type: types.GET_CLIENTS
    }
}

export const getClientsSuccess = (payload) => {
    return function(dispatch){
        dispatch({
             type: types.GET_CLIENTS_SUCCESS,
             payload
        });
     }   
}

export const getClientsFailure = (payload) => {
    return function(dispatch){
       dispatch({
            type: types.GET_CLIENTS_FAILURE,
            payload
       });
    }   
}

export const getClient = () => {
    return {
        type: types.GET_CLIENT
    }
}

export const getClientSuccess = (payload) => {
    return function(dispatch){
        dispatch({
             type: types.GET_CLIENT_SUCCESS,
             payload
        });
     }   
}

export const getClientFailure = (payload) => {
    return function(dispatch){
       dispatch({
            type: types.GET_CLIENT_FAILURE,
            payload
       });
    }   
}


export const getRequested = () => {
    return {
        type: types.GET_REQUESTED
    }
}

export const getRequestedSuccess = (payload) => {
    return {
        type: types.GET_REQUESTED_SUCCESS,
        payload
    }
}


export const getRequestedFailure = (payload) => {
    return {
        type: types.GET_REQUESTED_FAILURE,
        payload
    }
}

export const getApproved = () => {
    return {
        type: types.GET_APPROVED
    }
}

export const getApprovedSuccess = (payload) => {
    return {
        type: types.GET_APPROVED_SUCCESS,
        payload
    }
}


export const getApprovedFailure = (payload) => {
    return {
        type: types.GET_APPROVED_FAILURE,
        payload
    }
}

export const approveResources = () => {
    return {
        type: types.APPROVE_RESOURCES
    }
}

export const approveResourcesSuccess = (payload) => {
    return function(dispatch){
        dispatch({
             type: types.APPROVE_RESOURCES_SUCCESS,
             payload
        });
     }   
}

export const approveResourcesFailure = (payload) => {
    return function(dispatch){
       dispatch({
            type: types.APPROVE_RESOURCES_FAILURE,
            payload
       });
    }   
}

export const declineResources = () => {
    return {
        type: types.DECLINE_RESOURCES
    }
}

export const declineResourcesSuccess = (payload) => {
    return function(dispatch){
        dispatch({
             type: types.DECLINE_RESOURCES_SUCCESS,
             payload
        });
     }   
}

export const declineResourcesFailure = (payload) => {
    return function(dispatch){
       dispatch({
            type: types.DECLINE_RESOURCES_FAILURE,
            payload
       });
    }   
}

export const setClient = (payload) => {
    return function(dispatch){
       dispatch({
            type: types.SET_CLIENT,
            payload
       });
    }   
}