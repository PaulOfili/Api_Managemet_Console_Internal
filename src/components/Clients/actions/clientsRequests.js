import request from "../../../request/request"
import api from "../../../request/endpoints";


export const getProjects = () => { 
    let options = {
        ...api.getProjects(),
    };
    return request.make(options, true)
}

export const getProject = (projectId) => { 
    let options = {
        ...api.getProject(projectId),
    };
    return request.make(options, true)
}

export const getRequested = (projectId) => { 
    let options = {
        ...api.getRequested(projectId),
    };
    return request.make(options, true)
}

export const getApproved = (projectId) => { 
    let options = {
        ...api.getApproved(projectId),
    };
    return request.make(options, true)
}

export const approveResources = (body, projectId) => { 
    let data = body
    let options = {
        ...api.approveResources(projectId),
        data
    };
    return request.make(options, true)
}

export const declineResources = (body, projectId) => { 
    let data = body
    let options = {
        ...api.declineResources(projectId),
        data
    };
    return request.make(options, true)
}
