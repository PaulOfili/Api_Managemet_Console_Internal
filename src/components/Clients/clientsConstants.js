import keyMirror  from 'key-mirror';

export default keyMirror({
  SET_CLIENT: null,
  GET_CLIENTS: null,
  GET_CLIENTS_SUCCESS: null,
  GET_CLIENTS_FAILURE: null,
  GET_CLIENT: null,
  GET_CLIENT_SUCCESS: null,
  GET_CLIENT_FAILURE: null,
  GET_PASSPORT_CLIENT_SUCCESS: null,
  GET_PASSPORT_CLIENT_FAILURE: null,
  SAVE_CLIENT:null,
  SAVE_CLIENT_SUCCESS: null,
  SAVE_CLIENT_FAILURE: null,
  GET_REQUESTED: null,
  GET_REQUESTED_SUCCESS: null,
  GET_REQUESTED_FAILURE: null,
  GET_APPROVED: null,
  GET_APPROVED_SUCCESS: null,
  GET_APPROVED_FAILURE: null,
  APPROVE_RESOURCES: null,
  APPROVE_RESOURCES_SUCCESS: null,
  APPROVE_RESOURCES_FAILURE: null,
  DECLINE_RESOURCES: null,
  DECLINE_RESOURCES_SUCCESS: null,
  DECLINE_RESOURCES_FAILURE: null
});
