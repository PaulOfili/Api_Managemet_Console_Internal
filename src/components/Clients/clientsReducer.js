import types from './clientsConstants';

const initialState = {
    loading: false,
    clients: [],
    requested: [],
    approved: [],
    client: {}
}

export default (state = initialState, action) => {
    switch (action.type) {
        case types.SAVE_CLIENT:
            return (
                Object.assign({}, state, {
                    loading: true,
                })
            );
        case types.SAVE_CLIENT_SUCCESS:
            return (
                Object.assign({}, state, {
                    clients: action.payload,
                    clearForm: true,
                    loading: false,
                })
            );
        case types.SAVE_CLIENT_FAILURE:
            return (
                Object.assign({}, state, {
                    loading: false,
                })
            );
        case types.GET_CLIENTS:
            return (
                Object.assign({}, state, {
                    loading: true,
                })
            );
        case types.GET_CLIENTS_SUCCESS:
            return (
                Object.assign({}, state, {
                    clients: action.payload,
                    loading: false,
                })
            );
        case types.GET_CLIENTS_FAILURE:
            return (
                Object.assign({}, state, {
                    loading: false,
                })
            );
        case types.GET_CLIENT:
            return (
                Object.assign({}, state, {
                    loading: true,
                })
            );
        case types.GET_CLIENT_SUCCESS:
            return (
                Object.assign({}, state, {
                    client: action.payload,
                    loading: false,
                })
            );
        case types.GET_CLIENT_FAILURE:
            return (
                Object.assign({}, state, {
                    loading: false,
                })
            );
        case types.GET_REQUESTED:
            return (
                Object.assign({}, state, {
                    loading: true,
                })
            );
        case types.GET_REQUESTED_SUCCESS:
            return (
                Object.assign({}, state, {
                    requested: action.payload,
                    loading: false,
                })
            );
        case types.GET_REQUESTED_FAILURE:
            return (
                Object.assign({}, state, {
                    loading: false,
                    error: action.payload
                })
            );
        case types.GET_APPROVED:
            return (
                Object.assign({}, state, {
                    loading: true,
                })
            );
        case types.GET_APPROVED_SUCCESS:
            return (
                Object.assign({}, state, {
                    approved: action.payload,
                    loading: false,
                })
            );
        case types.GET_APPROVED_FAILURE:
            return (
                Object.assign({}, state, {
                    loading: false,
                    error: action.payload
                })
            );
        case types.SET_CLIENT:
            return (
                Object.assign({}, state, {
                    client: action.payload,
                    loading: false,
                })
            );
        default:
            return state;
    }
}