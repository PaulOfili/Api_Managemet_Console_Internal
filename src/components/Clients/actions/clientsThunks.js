// import mockprojectApi from '../../../request/mockData';
import * as projectsApi from './clientsRequests';
import * as actions from './clientsActions';
import { showLoading, hideLoading } from 'react-redux-loading-bar';
import { showAlert } from '../../../alert';

export const getClients = () => {
    return async dispatch => {
            dispatch(actions.getClients());
            dispatch(showLoading());
            try {
                const response = await projectsApi.getProjects();
                response && dispatch(actions.getClientsSuccess(response.data));
                dispatch(hideLoading());
            }
            catch(error){
                dispatch(hideLoading());
                if(error.data && error.data.message){

                    dispatch(showAlert({
                        type:'error',
                        message: error.data.message        
                    }, dispatch))
                }
                else {
                    dispatch(showAlert({
                        type:'error',
                        message: "Unable to get response from server"          
                    }, dispatch))
                }
                dispatch(actions.getClientsFailure(error));
            }
    }
}

export const getClient = (projectId) => {
    return async dispatch => {
            dispatch(actions.getClient(projectId));
            dispatch(showLoading());
            try {
                const response = await projectsApi.getProject(projectId);
                response && dispatch(actions.getClientSuccess(response.data));
                dispatch(hideLoading());
            }
            catch(error){
                dispatch(hideLoading());
                if(error.data && error.data.message){

                    dispatch(showAlert({
                        type:'error',
                        message: error.data.message        
                    }, dispatch))
                }
                else {
                    dispatch(showAlert({
                        type:'error',
                        message: "Unable to get response from server"          
                    }, dispatch))
                }
                dispatch(actions.getClientFailure(error));
            }
    }
}

export const getRequested = (projectId) => {
    return async dispatch => {
            dispatch(actions.getRequested());
            dispatch(showLoading());
            try {
                const response = await projectsApi.getRequested(projectId);
                response && dispatch(actions.getRequestedSuccess(response.data));
                dispatch(hideLoading());
            }
            catch(error){
                dispatch(hideLoading());
                if(error.data && error.data.message){

                    dispatch(showAlert({
                        type:'error',
                        message: `Unable to get response from server. ${error.data.message}.`     
                    }, dispatch))
                }
                else {
                    dispatch(showAlert({
                        type:'error',
                        message: "Unable to get response from server"          
                    }, dispatch))
                }
                dispatch(actions.getRequestedFailure(error));
            }
    }
}

export const getApproved = (projectId) => {
    return async dispatch => {
            dispatch(actions.getApproved());
            dispatch(showLoading());
            try {
                const response = await projectsApi.getApproved(projectId);
                response && dispatch(actions.getApprovedSuccess(response.data));
                dispatch(hideLoading());
            }
            catch(error){
                dispatch(hideLoading());
                if(error.data && error.data.message){

                    dispatch(showAlert({
                        type:'error',
                        message: `Unable to get response from server. ${error.data.message}.`       
                    }, dispatch))
                }
                else {
                    dispatch(showAlert({
                        type:'error',
                        message: "Unable to get response from server"          
                    }, dispatch))
                }
                dispatch(actions.getApprovedFailure(error));
            }
    }
}

export const setResourceList = (newObject) => {
    return dispatch => {
        dispatch(actions.setClient(newObject));
    }
}

export const approveResources = (body, projectId) => {
    return async dispatch => {
            dispatch(actions.approveResources());
            dispatch(showLoading());
            try {
                const response = await projectsApi.approveResources(body, projectId);
                response && dispatch(actions.approveResourcesSuccess(response.data));
                dispatch(getApproved(projectId));
                dispatch(getRequested(projectId));
                dispatch(setResourceList({}))
                dispatch(hideLoading());
            }
            catch(error){
                dispatch(hideLoading());
                if(error.data && error.data.message){

                    dispatch(showAlert({
                        type:'error',
                        message: `${error.data.message}. Note: The project must be live before approving resources.`      
                    }, dispatch))
                }
                else {
                    dispatch(showAlert({
                        type:'error',
                        message: "Unable to get response from server"          
                    }, dispatch))
                }
                dispatch(actions.approveResourcesFailure(error));
            }
    }
}

export const declineResources = (body, projectId) => {
    return async dispatch => {
            dispatch(actions.declineResources());
            dispatch(showLoading());
            try {
                const response = await projectsApi.declineResources(body, projectId);
                response && dispatch(actions.declineResourcesSuccess(response.data));
                dispatch(getApproved(projectId));
                dispatch(getRequested(projectId));
                dispatch(setResourceList({}))
                dispatch(hideLoading());
            }
            catch(error){
                dispatch(hideLoading());
                if(error.data && error.data.message){

                    dispatch(showAlert({
                        type:'error',
                        message: error.data.message        
                    }, dispatch))
                }
                else {
                    dispatch(showAlert({
                        type:'error',
                        message: "Unable to get response from server"          
                    }, dispatch))
                }
                dispatch(actions.declineResourcesFailure(error));
            }
    }
}
