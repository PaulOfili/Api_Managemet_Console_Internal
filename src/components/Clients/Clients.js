import React, { useEffect, useState } from 'react';
import ClientsTable from './ClientsTable';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as clientsRequests from "./actions/clientsThunks";
import { Input } from 'antd';

const { Search } = Input;

function Clients({ clients, getClients }){ 
    const [searchText, setSearchText] = useState('');
    const [clientListBySearch, setClientListBySearch] = useState([]);

    useEffect(() => {
        getClients();
    }, [getClients])

    useEffect(() => {
        setClientListBySearch([...clients])
    }, [clients])

    const handleChange = (e) => {
        setSearchText(e.target.value)
    }

    const onSearch = () => {
        const newClientList = clients.filter(client => client.name.includes(searchText));
        setClientListBySearch(newClientList);
    }

    const data = clientListBySearch.map((client, index) => {
        const {  id, name, type, owner } = client
        return {
            key: index,
            id,
            name,
            type,
            owner
        }
    })
    return (
        <section className='main-content'>
            <div className="sub-header">
                <h3>Projects</h3>
            </div>
            <section className='content-wrapper route_section'>
                <div className="splitroutes-text">
                    <div className="item">Here you can approve or decline resources in projects</div>
                    <div className="item">Note that only projects listed below are being used by developers.</div>
                </div>
                <div className='m-5'>
                    <Search placeholder="Search by Project/Client Name" size='large' onChange={handleChange} onSearch={onSearch} value={searchText} allowClear enterButton />
                </div>
                <section className='splitroutes-tablearea'>
                    <ClientsTable data={data}/>
                </section>
            </section>
        </section>
    )
}

const mapStateToProps = (state) => {
    return {
        clients: state.clients.clients,
        isLoading: state.clients.loading,
    }
}

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
        ...clientsRequests,
    }, dispatch)
  };

export { Clients }
export default connect(mapStateToProps, mapDispatchToProps)(Clients);