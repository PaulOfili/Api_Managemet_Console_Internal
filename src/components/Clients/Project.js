import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {Empty} from "antd";
import { Icon, Message } from 'semantic-ui-react';
import * as clientsRequests from "./actions/clientsThunks";
import emptyBoxSvg from '../../assets/img/empty-box.svg';
import Accordion from './Accordion';

function Project({ match, getRequested, getApproved, approved, requested, isLoading, approveResources, declineResources, client, setResourceList }){

    // const [approve, setApprove] = useState({});

    useEffect(() => {
        getRequested(match.params.id)
        getApproved(match.params.id)
    }, [match.params.id, getRequested, getApproved])

    const handleRadio = (e, { value, name }) => { 
        setResourceList({...client, [name]: value})
        // setApprove({...approve, [name]:value}) 
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        let approvedResources = [];
        let declinedResources = [];
        for (var key in client) {
            if (client.hasOwnProperty(key)) {
              if(client[key] === 'approve')
                approvedResources.push(key)
              if(client[key] === 'decline')
                declinedResources.push(key)
            }
        }
        
        if(approvedResources.length > 0)
            approveResources({ resources: approvedResources}, match.params.id);

        if(declinedResources.length > 0)
            declineResources({ resources: declinedResources}, match.params.id)
    }
    return (
        <section className='main-content'>
                <div className="sub-header">
                    <h3>{match.params.id}</h3>
                </div>
                <section className='route_section'>
                    <div className="splitroutes-text">
                        <div className="item">Here you can approve or decline resources in products for a project</div>
                        <div className="item">Note that only resources listed below are in use by developers.</div>
                    </div>
                    <section>
                        <Message info icon>
                            <Icon name='hand point down'  />
                            <Message.Content>
                                <Message.Header>REQUESTED</Message.Header>
                                <p>Select resources in Products and Approve, Decline or Neglect(remains in Pending state)</p>
                            </Message.Content>
                        </Message>
                        <div>
                        {
                            !isLoading && requested.length === 0 &&
                            <Empty
                                image={emptyBoxSvg}
                                imageStyle={{
                                    height: '3rem',
                                    width: '3rem',
                                    margin: '12.4375rem auto 1.5rem'
                                }}
                                description={
                                    <span>
                                        <span style={{lineHeight: '1.75rem', fontSize: '1.5rem', display: 'block'}}>
                                            <span>No Requested Resources for this project</span>
                                        </span>
                                        <span style={{lineHeight: '1.25rem', fontSize: '1rem', padding: '1rem',
                                            width: '21.0625rem',
                                            margin: '0 auto',
                                            display: 'block'}}>
                                            <span></span>
                                        </span>
                                    </span>
                                }
                            >
                            </Empty>
                            }
                            {
                                !isLoading && requested.length > 0 &&
                                <div>
                                    <Accordion handleRadio={handleRadio} array={requested} approve={client} type={'requested'}/>
                                    <button className='btn btn-primary mt-5 mb-5 right-float' onClick={handleSubmit}>Submit</button>
                                </div>
                            }
                            
                        </div>
                    </section>
                    <section>
                        <Message info icon>
                            <Icon name='hand point down'  />
                            <Message.Content>
                                <Message.Header>APPROVED RESOURCES</Message.Header>
                                <p>Find below all approved Resources for the current project</p>
                            </Message.Content>
                        </Message>
                        <div>
                            {
                               !isLoading && approved.length === 0 &&
                                <Empty
                                    image={emptyBoxSvg}
                                    imageStyle={{
                                        height: '3rem',
                                        width: '3rem',
                                        margin: '12.4375rem auto 1.5rem'
                                    }}
                                    description={
                                        <span>
                                            <span style={{lineHeight: '1.75rem', fontSize: '1.5rem', display: 'block'}}>
                                                <span>No Approved Resources for this project</span>
                                            </span>
                                            <span style={{lineHeight: '1.25rem', fontSize: '1rem', padding: '1rem',
                                                width: '21.0625rem',
                                                margin: '0 auto',
                                                display: 'block'}}>
                                                <span></span>
                                            </span>
                                        </span>
                                    }
                                >
                                </Empty>
                            }
                            {
                                !isLoading && approved.length > 0 &&
                                <Accordion handleRadio={handleRadio} array={approved} approve={false} type={'approved'}/>
                            }
                        </div>
                    </section>
                </section>
        </section>
    );


}

const mapStateToProps = (state) => {
    return {
        client: state.clients.client,
        requested: state.clients.requested,
        approved: state.clients.approved,
        isLoading: state.clients.loading,
    }
}

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
        ...clientsRequests,
    }, dispatch)
  };

export default connect(mapStateToProps, mapDispatchToProps)(Project);