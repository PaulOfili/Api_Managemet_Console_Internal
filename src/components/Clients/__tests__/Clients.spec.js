import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import {Clients} from "../Clients";
import * as clientRequests from '../actions/clientsThunks';

Enzyme.configure({ adapter: new Adapter() });

function setup() {
  const props = {
    clients: [],
    isLoading: false,
    dispatch: jest.fn(),
    ...clientRequests
  };
  const enzymeWrapper = shallow(<Clients {...props} />);
  return {
    props,
    enzymeWrapper
  };
}

describe("components", () => {
  describe("Clients", () => {
    it("To match snapshot", () => {
      const { enzymeWrapper } = setup();
      expect(enzymeWrapper).toMatchSnapshot();
    });
  });
});
