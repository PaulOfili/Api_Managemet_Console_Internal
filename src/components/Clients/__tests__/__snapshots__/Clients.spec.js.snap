// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Clients To match snapshot 1`] = `ShallowWrapper {}`;
